ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralDeserializer.java

```diff
@@ -0,0 +1,170 @@
+      columnTypes = new ArrayList<TypeInfo>();
+    } else {
+      columnTypes = TypeInfoUtils
+          .getTypeInfosFromTypeString(columnTypeProperty);
+    }
+
+    if (columnNames.size() != columnTypes.size()) {
+      throw new SerDeException("ThriftGeneralDeserializer initialization failed. Number of column " +
+          "name and column type differs. columnNames = " + columnNames + ", columnTypes = " +
+          columnTypes);
+    }
+
+    ThriftFieldIdResolver resolver = HiveThriftFieldIdResolverFactory.createResolver(tbl);
+    
+    List<ObjectInspector> columnObjectInspectors = new ArrayList<>();
+    for (int i = 0; i < columnNames.size(); i++)
+      columnObjectInspectors.add(getObjectInspector(columnTypes.get(i), resolver.getNestedResolver(i)));
+
+    cachedObjectInspector = new ThriftGeneralObjectInspector(columnNames, columnObjectInspectors, resolver);
+
+    // output debug info
+    LOG.debug("ThriftGeneralDeserializer initialized with: columnNames=" + columnNames
+        + " columnTypes=" + columnTypes);
+
+    stats = new SerDeStats();
+  }
+
+  private ObjectInspector getObjectInspector(TypeInfo info, ThriftFieldIdResolver resolver) {
+    ObjectInspector result = null;
+    switch (info.getCategory()) {
+      case PRIMITIVE:
+        result = PrimitiveObjectInspectorFactory.getPrimitiveJavaObjectInspector(TypeInfoFactory.getPrimitiveTypeInfo(info.getTypeName()));
+        break;
+      case LIST: {
+        ObjectInspector elementObjectInspector = getObjectInspector(((ListTypeInfo) info)
+            .getListElementTypeInfo(), resolver.getNestedResolver(0));
+        result = ObjectInspectorFactory.getStandardListObjectInspector(elementObjectInspector);
+        break;
+      }
+      case MAP: {
+        MapTypeInfo mapTypeInfo = (MapTypeInfo) info;
+        ObjectInspector keyObjectInspector = getObjectInspector(mapTypeInfo
+            .getMapKeyTypeInfo(), resolver.getNestedResolver(0));
+        ObjectInspector valueObjectInspector = getObjectInspector(mapTypeInfo
+            .getMapValueTypeInfo(), resolver.getNestedResolver(1));
+        result = ObjectInspectorFactory
+            .getStandardMapObjectInspector(keyObjectInspector,
+                valueObjectInspector);
+        break;
+      }
+      case STRUCT: {
+        StructTypeInfo structTypeInfo = (StructTypeInfo) info;
+        List<String> fieldNames = structTypeInfo.getAllStructFieldNames();
+        List<TypeInfo> fieldTypeInfos = structTypeInfo
+            .getAllStructFieldTypeInfos();
+        List<ObjectInspector> fieldObjectInspectors = new ArrayList<ObjectInspector>(
+            fieldTypeInfos.size());
+        for (int i = 0; i < fieldTypeInfos.size(); i++) {
+          fieldObjectInspectors
+              .add(getObjectInspector(fieldTypeInfos.get(i), resolver.getNestedResolver(i)));
+        }
+        result = new ThriftGeneralObjectInspector(fieldNames, fieldObjectInspectors, resolver);
+        break;
+      }
+      case UNION: {
```

Comment:
HIVE support C style union where it can contain data of any type but only one field at a time. I am not sure thrift supports something like that. So leaving it empty makes sense.